pilot/pkg/proxy/envoy/v1/watcher.go

```diff
@@ -208,6 +221,47 @@ func watchCerts(ctx context.Context, certsDirs []string, watchFileEventsFn watch
 	watchFileEventsFn(ctx, fw.Event, minDelay, updateFunc)
 }
 
+func certsExist(certs []CertSource, interval time.Duration, timeout time.Duration) error {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		for _, cert := range certs {
+			for {
+				if err := checkCerts(cert); err != nil {
```

Comment:
This goroutine will leak if the file never exists. Something like the following using `errgroup` and `context` packages might work. The total time is bounded by the context timeout (e.g. 10 seconds) and the individual file existence checks are performed in parallel with 1 second retry. 

```golang
func exists(files []string) error {
    ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
    g, ctx := errgroup.WithContext(ctx)
    for i := range files {
        filename := files[i]
        g.Go(func() error {
            for {
                if _, err := os.Stat(filename); !os.IsNotExist(err) {
                    return nil
                }
                select {
                case <-time.After(time.Second):
                    // try again                                                                                                                                     
                case <-ctx.Done():
                    return fmt.Errorf("failed to read %v", filename)
                }
            }
        })
    }
    return g.Wait()
}
```

Reply:
Thanks. I have updated the code using the errgroup approach.